### ISSUE TITLE
Not possible to rearrange tabs by dragging

### ISSUE THREAD

**Issue by johanbrook** [zeit/hyper#39]
The whole Hyperterm window is moved when dragging a tab with the mouse.

**Version:** 0.2.1

_(I'm filing a bunch of UI stuff now, to keep track of them. Can be triaged)._

**Comment by rauchg** [comment]
Yep. This is gonna be a tricky one. I'm thinking about a "3d touch" / prolonged click and hold type thing where the tab will become manageable to drag around.

I believe that if we constrain ourselves to the HTML5 API, it should be possible to "receive" it in other windows in fact :)

**Comment by numeraltwo** [comment]
Might be some learnings in here: https://github.com/atom/tabs

**Comment by Qard** [comment]
Yeah, I would have a short delay and then have the grabbed tab visual pop out somehow to indicate to the user that they are dragging the tab now.

**Comment by rhys-vdw** [comment]
Is there already a command to do this via keyboard?

**Comment by patrik-piskay** [comment]
I've just published a plugin for tabs drag&drop - https://github.com/patrik-piskay/hyperterm-tabs

Unfortunately, I had to set the title bar to be visible as I wasn't able to make it work without it (and still have the possibility of moving term window around). I've tried the approach of a long click that would make the terminal window not movable (and tabs drag&droppable) but it wouldn't take effect until after mouse button was released - probably an OS constraint but if somebody will find/know a solution for it, let me know!

**Comment by nkkollaw** [comment]
Fixing this should be a priority.

I went back to iTerm just because of being able to rearrange tabs, which I do often.

Otherwise the project is awesome.

**Comment by chabou** [comment]
[hyper-term-tabs](https://github.com/patrik-piskay/hyperterm-tabs) has this feature.

**Comment by patrik-piskay** [comment]
Hi @chabou, it used to have this feature but Hyper v0.8 made it not possible anymore. I've opened an [issue](https://github.com/zeit/hyper/issues/911) for that in this repo. So currently `hyperterm-tabs` support tabs reordering by keyboard shortcuts only, no drag&drop at the moment.

**Comment by chabou** [comment]
Ok.
Maybe it should be a core feature

**Comment by chabou** [comment]
Problem is: how to have tab dragging and not impacting window dragging.
Add margin is not ideal.
Maybe enable tab dragging after a long (~500ms) press?

**Comment by nkkollaw** [comment]
It might make sense exploring the -webkit-user-drag property.

If I understand how it works, if you have:

```
<ul class="tab-bar">
    <li class="tab">Tab 1</li>
    <li class="tab">Tab 2</li>
</ul>
```

And you do:

```
.tab-bar {
    -webkit-user-drag: .tab
}
```

Dragging would work on the tab instead of the tab bar.

I don't know much else about it, though. Is there someone smarter than me that can give it a try?

**Comment by philip-peterson** [comment]
Might as well toss an opinion in here: maybe option-click could reorganize tabs, while regular click just moves the window?

**Comment by nkkollaw** [comment]
Very nice. I personally rarely reorganize them, but it's frustrating not being able to do it when I do want to do it.

**Comment by mischah** [comment]
@chabou

> Problem is: how to have tab dragging and not impacting window dragging.
> Add margin is not ideal.
> Maybe enable tab dragging after a long (~500ms) press?

You can find possible solutions over here:
https://github.com/zeit/hyper/issues/911#issuecomment-333417354

**Comment by nwoolls** [comment]
Check out Firefox. It has tabs in the chrome, but there's still room on either side of the tabs to grab the window itself.

**Comment by nkkollaw** [comment]
It's really unbelievable that you guys are still talking about this, like Hyper was the first software to use tabs.

What is there to talk about? Either Hyper devs want to support this feature, or not.

**Comment by nkkollaw** [comment]
I think this is a perfect example of how inefficient FOSS is when there is a decision to make.

It's tabs. 

- every browsers has tabs
- every terminal in existence has tabs
- file managers have tabs
- Hyper: "mmm... how should tabs work? Wow can we both have tabs and drag the window at the same time? Let's think about for a few years

**Comment by andradei** [comment]
Haven't touched Hyper since 2016 mainly for this reason. I could use `tmux` to solve this, but meh. iTerm is a great terminal.

This is a great project, however, a terminal powered by web tech can do amazing things.

**Comment by nkkollaw** [comment]
@andradei : same here.

I tried again yesterday to see if it was fixed, I'm just posting here because of that. It's really a pity, because otherwise it looks really good.

I'll try again in 2020 or something.

**Comment by knownasilya** [comment]
Well it is marked 'help wanted', meaning they'll accept a pull request, and since this is OSS, the people that want it most should do the work when possible. The maintainer(s) is not obligated to fulfill every request in x amount of time, or ever.

**Comment by nkkollaw** [comment]
This is a legend, @knownasilya.

First of all, a huge number FOSS projects are sponsored by a big company that pays people to develop it.

Even if this is not the case, it's not a democratic process. If the devs don't want to do something, they won't merge the merge request, and the only way to do it is forking. That's why you have so many forked projects.

**Comment by jocull** [comment]
Not to pile-in on the discussion too much, but there's a huge number of unresolved or stuck PRs. There seems to be quite a bit of churn in making concrete progress, exacerbated by a lack of automated testing (and thus confidence) in some key areas. These aren't easy problems to solve, but it's definitely bigger than just opening PRs :smile:

**Comment by andradei** [comment]
@knownasilya I agree the maintainers don't have any obligation to tackle this issue, I'm not mad at them. I'm rather amazed at the project actually. I also want this feature and wish I had the time to learn how to contribute. It is clear that the maintainers have other priorities and that's how it should be. So I'll check next year to see if the feature has been implemented.

**Comment by nkkollaw** [comment]
What's going on with this political correctness, @andradei? Everyone knows that FOSS devs are not under any obligation, we don't have to compliment them every time an issue is opened.

The point is just: is this "WONTFIX", or not. It's just crazy to keep talking about how tabs should work when hundred of apps have been using tabbed content for years.

**Comment by andradei** [comment]
@nkkollaw I agree with you about

> The point is just: is this "WONTFIX", or not. It's just crazy to keep talking about how tabs should work when hundred of apps have been using tabbed content for years.

But disagree with

> What's going on with this political correctness, @andradei? Everyone knows that FOSS devs are not under any obligation, we don't have to compliment them every time an issue is opened.

I'm not being politically correct. I've stated the issue and my frustration while deliberately showing gratitude for this project existing. I don't compliment them because I have to, but because I want to. Even if that's not directly related to the issue.

**Comment by nkkollaw** [comment]
@andradei 

> I'm not being politically correct. I've stated the issue and my frustration while deliberately showing gratitude for this project existing. I don't compliment them because I have to, but because I want to. Even if that's not directly related to the issue.

Jesus Christ, man! Whatever.

**Comment by hegelstad** [comment]
Is this labelled "Wont fix" by the maintainers, I am wondering because I have my hopes up?

**Comment by saamalik** [comment]
Just to add a little _spice_, my first tab _somehow_ moved to be tab _3_. I can't for the life of me reproduce it again, but hey - maybe some bugs are features!

**Comment by ppot** [comment]
@saamalik 😂 You fond the easter 🥚 !!! 🎉 . Look like a renderer issue somehow.

**Comment by andradei** [comment]
So this is labeled "Feature Request" since I don't know when. But by now I learned how to use tmux  which totally takes the need of tabs away. Yet, this feature is nice to have for those who don't like the make keyboard commands to get windows and panes opened.

**Comment by JZ-at-TP** [comment]
I've stopped using the app with this missing, its just not usable without. If it's ever completed I'll look at it again.

**Comment by realies** [comment]
three years later tabs are still not draggable?

**Comment by vuther** [comment]
If Hyper could drag tabs, I'd switch over to it too. It's honestly unbelievable that after 3 years, this feature isn't implemented. Seems like the devs are on cruise control right now.

**Comment by saamalik** [comment]
You guys do realize this is an open-source project? Feel free to implement it.

**Comment by aachman22** [comment]
> Just to add a little _spice_, my first tab _somehow_ moved to be tab _3_. I can't for the life of me reproduce it again, but hey - maybe some bugs are features!

This happened to me today too, which that prompted me to do a google search to check if this issue had been resolved and they had mapped a keyboard shortcut for changing tabs. Seems like a one-off case :/

**Comment by schkolne** [comment]
FWIW this is the only problem I've had with Hyper after using it for about 6months.  I ain't gonna swap to iTerm or anything Hyper is just too pretty to walk away from... but -- now that Mac has the ability to make *everything* tabbable, detachable, gather-able, and of course you can rearrange... this feels like a glaring omission.  +1 to fixing this

**Comment by bswank** [comment]
So, this is fixed now?

**Comment by romaklimenko** [comment]
Hello from 2020. Sorry to see that this feature is not going to happen. It forces me to switch back to Terminal.

**Comment by Robbie-Cook** [comment]
This be some bullsheet right here. This could be the best terminal, but instead, it's feature lacking.

**Comment by philip-peterson** [comment]
Maybe time to hook up hyper to a bug bounty system so issues like this can be funded.

**Comment by cherryblossom000** [comment]
I switched to [iTerm2](https://iterm2.com/) because of this issue. iTerm2 is a lot more powerful than Hyper and I recommend it for macOS users.

**Comment by drale2k** [comment]
Open source or not, you just gotta have some features and not being able to sort tabs is bonkers. Hope this gets fixed sometime

**Comment by etos** [comment]
madness

**Comment by philip-peterson** [comment]
anyone got a bounty for this bug?
